### PR TITLE
WIP: add slint::Window::(set_)position

### LINKED ISSUE #323: Add API to control the position of the window on the screen
Currently the public API in Rust/C++/JavaScript only allows showing or hiding the window. This issue tracks the availability of API to control the position of the window on the screen, towards the windowing system / window manager.

### PR DESCRIPTION
**Pull request by tronical**
TODO:
 - C++ API
 - JavaScript API
 - Check what types to re-export (or modules, like euclid?)

Fixes #323

### CODE REVIEW

File: internal/backends/qt/qt_window.rs [hunk 1]
```diff
@@ -1751,6 +1752,22 @@ impl PlatformWindow for QtWindow {
...
+
+    fn position(&self) -> euclid::Point2D<f32, PhysicalPx> {
+        let widget_ptr = self.widget_ptr();
+        let qp = cpp! {unsafe [widget_ptr as "QWidget*"] -> qttypes::QPoint as "QPoint" {
+            return widget_ptr->pos();
+        }};
+        euclid::Point2D::new(qp.x as _, qp.y as _)
+    }
+
+    fn set_position(&self, position: euclid::Point2D<f32, PhysicalPx>) {
```
Comment (ogoffart) [inline internal/backends/qt/qt_window.rs:1764]: The question is whether we want to use physical it logical here. Especially because this might be at a point where the scale factor is unknown

Comment (tronical) [reply internal/backends/qt/qt_window.rs]: Yes, that caused it not to work with the GL backend - I noticed that last night. That's fixed now :)

File: internal/core/api.rs [hunk 2]
```diff
@@ -11,6 +11,9 @@ use alloc::rc::Rc;
 use crate::component::ComponentVTable;
 use crate::window::WindowRc;
 
+pub use crate::lengths::LogicalPx;
+pub use crate::lengths::PhysicalPx;
```
[the remark this answers is not in the document]
Comment (tronical) [reply internal/core/api.rs:15]: We discussed this in person and concluded on aliases for now.

File: internal/core/api.rs [hunk 3]
```diff
@@ -11,6 +11,20 @@ use alloc::rc::Rc;
...
+/// Convenience alias for representing a size in logical pixels (before applying a window scale factor).
+pub type LogicalSize = euclid::Size2D<f32, LogicalPx>;
+/// Convenience alias for representing a size in physical pixels.
+pub type PhysicalSize = euclid::Size2D<i32, PhysicalPx>;
+/// Convenience alias for representing a point in logical pixels (before applying a window scale factor).
+pub type LogicalPoint = euclid::Point2D<f32, LogicalPx>;
+/// Convenience alias for representing a point in physical pixels.
+pub type PhysicalPoint = euclid::Point2D<i32, PhysicalPx>;
+/// Convenience alias for a scale that can be used with euclid's types to convert between logical and physical pixels.
+pub type LogicalToPhysicalScale = euclid::Scale<f32, LogicalPx, PhysicalPx>;
```
Comment (tronical) [inline internal/core/api.rs:26]: I wonder if `f32` and `i32` are the correct primitive data types here. i32 makes sense as physical pixel unit, but it means that after applying a the scale the result is an Euclid Point/Size that's `f32`, so conversion to an actual PhysicalPoint/PhysicalSize alias requires a `.cast()`.

Another option would be to say that the logical coordinate grid is also pixels in the API. But I guess that means scale factors would have to be integral values.

And finally, using `f32` everywhere is also a bit weird - in the light of actual pixels on the screen being discrete.

Comment (ogoffart) [reply internal/core/api.rs:26]: I think it sucks that rust can't do basic arithmetic between different type like integer and floats.

Now, what i'm concerned about is that the software renderer uses u16 for physical lenght and now we have two PhysicalLenghth.

I wonder if it is not simpler to just keep the type as `euclid::Point2D<i32, PhysicalPx>` then it's clear what it is.

But yes, that means we need to re-export euclid in a module.

File: internal/core/api.rs [hunk 4]
```diff
@@ -11,6 +11,20 @@ use alloc::rc::Rc;
 use crate::component::ComponentVTable;
 use crate::window::WindowRc;
 
+pub use crate::lengths::LogicalPx;
+pub use crate::lengths::PhysicalPx;
+
+/// Convenience alias for representing a size in logical pixels (before applying a window scale factor).
+pub type LogicalSize = euclid::Size2D<f32, LogicalPx>;
+/// Convenience alias for representing a size in physical pixels.
+pub type PhysicalSize = euclid::Size2D<i32, PhysicalPx>;
```
Comment (tronical) [inline internal/core/api.rs:20]: I also wonder if this should be `u32` instead of `i32`. Hmm.

File: internal/core/api.rs (hunk 3 above)
Comment (tronical) [reply internal/core/api.rs:26]: We could do that, but we also make our API less expressive then. Now:

`app.window().set_position(slint::PhysicalPoint::new(.., ..));`

and without the alias:

`app.window().set_position(slint::euclid::Point2D::new(.., ..));`

The tag is inferred.

That said, we can't make everything totally water-tight. The expected type needs to be documented with `set_position` anyway and the tag is definitely there when people start to store these values in variables. And then the type "safety" kicks in.

Comment (tronical) [reply internal/core/api.rs]: (Done)

**Review by ogoffart — APPROVED** [review]

File: internal/backends/testing/Cargo.toml [hunk 5]
```diff
@@ -19,3 +19,4 @@ path = "lib.rs"
 i-slint-core = { version = "=0.2.5", path = "../../../internal/core" }
 
 image = { version = "0.24.0", default-features = false, features = ["png", "jpeg"] }
+euclid = "0.22.1"
```
Comment (ogoffart) [inline internal/backends/testing/Cargo.toml:22]: do we need it, or can we not use the re-export?

Comment (tronical) [reply internal/backends/testing/Cargo.toml]: Good point, we don't need it. I've changed it also with all the other crates.

**Comment by tronical** [comment]
So on the coordinate side, winit is explicit in the API by taking a physical point and that works. `QWidget::setPosition()` takes "logical" coordinates.

File: api/cpp/include/slint_point.h [hunk 6]
```diff
@@ -0,0 +1,22 @@
+// Copyright © SixtyFPS GmbH <info@slint-ui.com>
```
Comment (ogoffart) [inline api/cpp/include/slint_point.h:1]: TBH i'm not a fan of adding one file per class like that (especially for small classes)
However, if you want to go that route, this means that we should split lots of our class  that are now in slint.h in different files

File: internal/core/window.rs [hunk 7]
```diff
@@ -908,4 +924,50 @@ pub mod ffi {
...
         window.request_redraw();
     }
+
+    /// Returns the position of the window on the screen, in physical screen coordinates and including
+    /// a window frame (if present).
+    #[no_mangle]
+    pub unsafe extern "C" fn slint_windowrc_position(
+        handle: *const WindowRcOpaque,
+        x: &mut i32,
+        y: &mut i32,
```
[the remark this answers is not in the document]
Comment (ogoffart) [reply internal/core/window.rs:934]: > despite it being seemingly declared.

We manage to do it for `slint_image_size`
Are you sure that Point2D is declared.  I assumed that you did the same trick as what is in slint_size.h to have a `template using` in the cbindgen_private namespace?

**Review by ogoffart — APPROVED** [review]

File: internal/backends/qt/qt_widgets/spinbox.rs [hunk 8]
```diff
@@ -1,6 +1,7 @@
 // Copyright © SixtyFPS GmbH <info@slint-ui.com>
 // SPDX-License-Identifier: GPL-3.0-only OR LicenseRef-Slint-commercial
 
+use i_slint_core::api::euclid;
 use i_slint_core::input::{FocusEventResult, KeyEventType};
 
 use super::*;
```
[the remark this answers is not in the document]
Comment (tronical) [reply internal/backends/qt/qt_widgets/spinbox.rs:7]: Indeed. I somehow overlooked that. I'll try to put it into qt_widgets.rs - that's less noisy.

File: api/cpp/include/slint_point.h (hunk 6 above)
Comment (tronical) [reply api/cpp/include/slint_point.h:1]: I'm not a fan either - I'm fine with a bigger file. I saw that we have `slint_size.h` and went for this.

I could make a ticket for the next release to consolidate: Move perhaps model and window into separate files and move simple data types like size and point into slint.h. What do you think?

File: internal/core/window.rs (hunk 7 above)
[the remark this answers is not in the document]
Comment (tronical) [reply internal/core/window.rs:934]: Yeah, that's what's puzzling me. I think I did everything the same. Let me try again and I can push that on top into a separate change - maybe you can spot the mistake :)

**Comment by tronical** [comment]
So with

```patch
diff --git a/api/cpp/include/slint.h b/api/cpp/include/slint.h
index 81080270f..598c2dbb8 100644
--- a/api/cpp/include/slint.h
+++ b/api/cpp/include/slint.h
@@ -175,12 +175,7 @@ public:

     void request_redraw() const { cbindgen_private::slint_windowrc_request_redraw(&inner); }

-    slint::Point<int> position() const
-    {
-        slint::Point<int> pos { 0, 0 };
-        cbindgen_private::slint_windowrc_position(&inner, &pos);
-        return pos;
-    }
+    slint::Point<int> position() const { return cbindgen_private::slint_windowrc_position(&inner); }

     void set_position(const slint::Point<int> &pos)
     {
diff --git a/internal/core/window.rs b/internal/core/window.rs
index 4067d3e75..2f6890fde 100644
--- a/internal/core/window.rs
+++ b/internal/core/window.rs
@@ -930,10 +930,9 @@ fn call(&self) -> CloseRequestResponse {
     #[no_mangle]
     pub unsafe extern "C" fn slint_windowrc_position(
         handle: *const WindowRcOpaque,
-        pos: &mut euclid::default::Point2D<i32>,
-    ) {
+    ) -> euclid::default::Point2D<i32> {
         let window = &*(handle as *const WindowRc);
-        *pos = window.position().to_untyped()
+        window.position().to_untyped()
     }

     /// Sets the position of the window on the screen, in physical screen coordinates and including
```

applied I get:

```sh
[4/20] Building CXX object examples/printerdemo/cpp_interpreted/CMakeFiles/printerdemo_interpreted.dir/main.cpp.o
In file included from /Users/simon/src/slint/examples/printerdemo/cpp_interpreted/main.cpp:4:
In file included from /Users/simon/src/slint/api/cpp/include/slint_interpreter.h:6:
In file included from /Users/simon/src/slint/api/cpp/include/slint.h:27:
/Users/simon/src/cppbuild/api/cpp/generated_include/slint_internal.h:1101:18: warning: 'slint_windowrc_position' has C-linkage specified, but returns incomplete type 'Point2D<int32_t>' (aka 'Point<int>') which could be incompatible with C [-Wreturn-type-c-linkage]
Point2D<int32_t> slint_windowrc_position(const WindowRcOpaque *handle);
                 ^
1 warning generated.
```

The thing is, this error message makes sense to me: We're evidently returning a C++ type from a function with C linkage. I don't quite understand why it works for `slint_image_size`.
